lib/internal/fs/copy/copy-sync.js

```diff
@@ -0,0 +1,340 @@
+  return getStats(destStat, src, dest, opts);
+}
+
+function getStats(destStat, src, dest, opts) {
+  const statSyncFn = opts.dereference ? statSync : lstatSync;
+  const srcStat = statSyncFn(src);
+
+  if (srcStat.isDirectory()) return onDir(srcStat, destStat, src, dest, opts);
+  else if (srcStat.isFile() ||
+           srcStat.isCharacterDevice() ||
+           srcStat.isBlockDevice()) {
+    return onFile(srcStat, destStat, src, dest, opts);
+  } else if (srcStat.isSymbolicLink()) {
+    return onLink(destStat, src, dest, opts);
+  } else if (srcStat.isSocket()) {
+    throw new ERR_FS_COPY_SOCKET({
+      code: 'EINVAL',
+      message: `cannot copy a socket file: ${dest}`,
+      path: dest,
+      syscall: 'copy',
+      errno: EINVAL,
+    });
+  } else if (srcStat.isFIFO()) {
+    throw new ERR_FS_COPY_FIFO_PIPE({
+      code: 'EINVAL',
+      message: `cannot copy a FIFO pipe: ${dest}`,
+      path: dest,
+      syscall: 'copy',
+      errno: EINVAL,
+    });
+  }
+  throw new ERR_FS_COPY_UNKNOWN({
+    code: 'EINVAL',
+    message: `cannot copy a unknown file type: ${dest}`,
+    path: dest,
+    syscall: 'copy',
+    errno: EINVAL,
+  });
+}
+
+function onFile(srcStat, destStat, src, dest, opts) {
+  if (!destStat) return copyFile(srcStat, src, dest, opts);
+  return mayCopyFile(srcStat, src, dest, opts);
+}
+
+function mayCopyFile(srcStat, src, dest, opts) {
+  if (opts.force) {
+    unlinkSync(dest);
+    return copyFile(srcStat, src, dest, opts);
+  } else if (opts.errorOnExist) {
+    throw new ERR_FS_COPY_EEXIST({
+      code: 'EEXIST',
+      message: `${dest} already exists`,
+      path: dest,
+      syscall: 'copy',
+      errno: EEXIST,
+    });
+  }
+}
+
+function copyFile(srcStat, src, dest, opts) {
+  copyFileSync(src, dest);
+  if (opts.preserveTimestamps) handleTimestamps(srcStat.mode, src, dest);
+  return setDestMode(dest, srcStat.mode);
+}
+
+function handleTimestamps(srcMode, src, dest) {
+  // Make sure the file is writable before setting the timestamp
+  // otherwise open fails with EPERM when invoked with 'r+'
+  // (through utimes call)
+  if (fileIsNotWritable(srcMode)) makeFileWritable(dest, srcMode);
+  return setDestTimestamps(src, dest);
+}
+
+function fileIsNotWritable(srcMode) {
+  return (srcMode & 0o200) === 0;
+}
+
+function makeFileWritable(dest, srcMode) {
+  return setDestMode(dest, srcMode | 0o200);
+}
+
+function setDestMode(dest, srcMode) {
+  return chmodSync(dest, srcMode);
+}
+
+function setDestTimestamps(src, dest) {
+  // The initial srcStat.atime cannot be trusted
+  // because it is modified by the read(2) system call
+  // (See https://nodejs.org/api/fs.html#fs_stat_time_values)
+  const updatedSrcStat = statSync(src);
+  return utimesMillisSync(dest, updatedSrcStat.atime, updatedSrcStat.mtime);
+}
+
+function utimesMillisSync(path, atime, mtime) {
+  const fd = openSync(path, 'r+');
+  futimesSync(fd, atime, mtime);
+  return closeSync(fd);
```

Comment:
Any reason this is using `futimes` instead of `utimes` (same question for the async version)? Can't this be simplified to
```suggestion
  return utimesSync(path, atime, mtime);
```